tests/lib_unit_test.c

```diff
@@ -0,0 +1,318 @@
+}
+END_TEST
+
+
+START_TEST(test_toggle_slots)
+{
+	// Test toggling slots
+	struct led_slot_list *sl = NULL;
+	bool devices_found = false;
+	led_status_t status = led_slots_get(ctx, &sl);
+
+	ck_assert_msg(status == LED_STATUS_SUCCESS, "led_slots_get %d", status);
+	if (status == LED_STATUS_SUCCESS) {
+		struct led_slot_list_entry *se;
+
+		while ((se = led_slot_next(sl))) {
+			if (led_controller_slot_support(led_slot_cntrl(se)) &&
+				slot_usable(led_slot_id(se))) {
+				enum led_ibpi_pattern led = led_slot_state(se);
+
+				led = (led == LED_IBPI_PATTERN_NORMAL) ?
+					LED_IBPI_PATTERN_LOCATE : LED_IBPI_PATTERN_NORMAL;
+				status = led_slot_set(ctx, se, led);
+				devices_found = true;
+				ck_assert_msg(status == LED_STATUS_SUCCESS,
+						"led_slot_set %d", status);
+				enum led_ibpi_pattern after_set = led_slot_state(se);
+
+				ck_assert_msg(led == after_set,
+						"%s led_slot_state expected = (%d) != actual (%d)",
+						led_slot_id(se), led, after_set);
+			}
+		}
+		led_slot_list_free(sl);
+	}
+	ck_assert_msg(devices_found, "No test LED devices found!");
+}
+END_TEST
+
+
+START_TEST(test_led_by_path)
+{
+	// Test led using device node path
+	struct led_slot_list *sl = NULL;
+	bool devices_found = false;
+	led_status_t status = led_slots_get(ctx, &sl);
+
+	ck_assert_msg(status == LED_STATUS_SUCCESS, "led_slots_get %d", status);
+	if (status == LED_STATUS_SUCCESS) {
+		struct led_slot_list_entry *se;
+
+		while ((se = led_slot_next(sl))) {
+			const char *device_node = led_slot_device(se);
+
+			if (led_controller_slot_support(led_slot_cntrl(se)) &&
+				slot_usable(led_slot_id(se)) && strlen(device_node)) {
+				char normalized[PATH_MAX];
+				enum led_ibpi_pattern led_via_slot;
+				enum led_ibpi_pattern led_via_path;
+				enum led_ibpi_pattern expected;
+
+				status = led_device_name_lookup(device_node, normalized);
+				ck_assert_msg(status == LED_STATUS_SUCCESS,
+						"led_device_name_lookup %d", status);
+				if (status != LED_STATUS_SUCCESS)
+					return;
+
+				devices_found = true;
+
+				led_via_slot = led_slot_state(se);
+				status = led_get(ctx, normalized, &led_via_path);
+				ck_assert_msg(status == LED_STATUS_SUCCESS, "led_get %d", status);
+				if (status != LED_STATUS_SUCCESS)
+					return;
+
+				ck_assert_msg(led_via_slot == led_via_path,
+						"slot led retrival %d != %d path led retrival",
+						led_via_slot, led_via_path);
+
+				expected = (led_via_path == LED_IBPI_PATTERN_NORMAL) ?
+					LED_IBPI_PATTERN_LOCATE : LED_IBPI_PATTERN_NORMAL;
+
+
+				status = led_set(ctx, normalized, expected);
+				ck_assert_msg(status == LED_STATUS_SUCCESS, "led_set %d", status);
+				if (status != LED_STATUS_SUCCESS)
+					return;
+
+				led_flush(ctx);
+
+				status = led_get(ctx, normalized, &led_via_path);
+				ck_assert_msg(status == LED_STATUS_SUCCESS, "led_get %d", status);
+				if (status != LED_STATUS_SUCCESS)
+					return;
+
+				led_via_slot = led_slot_state(se);
```

Comment:
> Those two controller applies for some LEDs. `set_led` chooses NPEM automatically- expected. We should respect controller priority that is why initially I proposed `get_default_controller()` to give library user possibility to check which controller should be used for slot manipulation to avoid corner cases.

A normal user of the library isn't going to know the intricacies of NPEM/VMD and which controller is the correct one to use.  IMHO, the library needs to handle this internally and do the right thing as does the command line.   If they use the API to select a device by device node path eg. `/dev/nvme2n1` or via slot `10000:02:07.0` it should just work. 

 What would we need to do in this case to simply not output VMD devices when NPEM is the only one that will work?

Reply:
>get_led relies on device->ibpi which must be set on set_led this is the one way to keep consistency, however I do not test how the change will affect ledmon (it seems that it must be tested, @bkucman- to notify), I'm surprised that it passed for you in this form.

It would be expected that a library user, once they have run `led_scan` that they should be able to retrieve the state of the LED without having to `set` it first.  I had to make changes to the SES path, so that `device->ibpi` gets set during the scan.  If there is an easier way to accomplish then, I'm all for it.

If we can't provide this consistency, then I suggest we remove the `led_get` and require using the slot functions to retrieve state and only provide `led_set` for hardware that doesn't have slot support.

Reply:
@mtkaczyk 

>They may both work. It is not forbidden, it is more like undefined behavior. I see excluding VMD as a worse solution because I'm using it for debugging purposes. It gives complete view.
I also know that comparting empty slots for both VMD/NPEM won't be simple.

You mentioned that `set_led` chooses NPEM automatically.  How is this decision made?  Does VMD only LED control exist?  We could provide a parameter to the `led_slots_get` function, something like a bit field (so we could easily add additional behavior switches in future) which allows the caller to specify that they want to retrieve all possible slots, even if they are non-functional/redundant?  We could then provide a switch to ledctl to set this value for debug. 

>There is no way because ledmon and ledctl were not created to provide get_led() functionality. It comes with get_slot.

The LED support we have in libStorageMgmt has the ability to retrieve LED status via device node.  I would like to be able to maintain that, so that I can use this library for it.  We already have `led_slot_find_by_device_name` so I think I can simply use that to retrieve the slot which will allow me to retrieve the current LED status.  So that allows `get_led` to be removed.

There is one thing that we are not handling though.  The SES leds we have in our lab have both an ID LED and a fault LED which can be set independently.  We don't have a way to express that today with our enumeration.  Suggestions?  For libStorageMgmt we use a bitfield, so that we can express status for both in one value.

Reply:
> You mentioned that `set_led` chooses NPEM automatically. How is this decision made? Does VMD only LED control exist? 

I'm referring to non-library code but I shouldn't change.

During `sysfs_scan` the `send_fn` is assigned to block device. ledmon and ledctl in non-slot mode uses send_fn function.
https://github.com/intel/ledmon/blob/master/src/block.c#L268
The decision is made basing on controller, and here we have a code:
https://github.com/intel/ledmon/blob/master/src/block.c#L187C26-L187C26
which promotes NPEM (that is how I implemented it).

>  We could provide a parameter to the `led_slots_get` function, something like a bit field (so we could easily add additional behavior switches in future) which allows the caller to specify that they want to retrieve all possible slots, even if they are non-functional/redundant? We could then provide a switch to ledctl to set this value for debug

Make sense to me. That will also require to change returning type to list, right? It will be a nice room for improvements.
It applies to set_slot by devnode so probably we will need to make 2 function `led_get_slot_by_devnode` and `led_get_slot_by_slot_id`- for slot_id controller is know so it will always return one slot.

It can also apply to set_slot, for example we can impose `locate` to every controller which supports this devnode using the flag. What do you think?
 
> > There is no way because ledmon and ledctl were not created to provide get_led() functionality. It comes with get_slot.

> 
> The LED support we have in libStorageMgmt has the ability to retrieve LED status via device node. I would like to be able to maintain that, so that I can use this library for it. We already have `led_slot_find_by_device_name` so I think I can simply use that to retrieve the slot which will allow me to retrieve the current LED status. So that allows `get_led` to be removed.

Excellent, but I didn't check if ledmon uses it, please verify.

> There is one thing that we are not handling though. The SES leds we have in our lab have both an ID LED and a fault LED which can be set independently. We don't have a way to express that today with our enumeration. Suggestions? For libStorageMgmt we use a bitfield, so that we can express status for both in one value.

I know, it is not supported right now. We should consider multiple patterns support in the future. Maybe we should return list instead on enum for get_slot method too?

I know that we are retuning values via pointers in library, so please do not take this "returning" directly.

Reply:
Today the library API has a call to retrieve the slots:

```c
led_status_t LED_SYM_PUBLIC led_slots_get(struct led_ctx *ctx, struct led_slot_list **sl);
```
You simply get all of them.  Ledctl is the one filtering.  I'm suggesting that we filter the list in the default case (add argument to change behavior) so that a user never gets the VMD slots or any slots returned to them that they cannot use.  At the moment I don't know how to provide that filtering as I don't understand the hardware and when it would work and when it wouldn't.  Can you explain this a bit more?

> Make sense to me. That will also require to change returning type to list, right?

Not exactly sure what you're referring too.  Today the library interface returns the following information about **each** slot:

* The device node, if it exists
* The slot identifer
* The controller
* The led_ibpi_pattern

>It applies to set_slot by devnode so probably we will need to make 2 function led_get_slot_by_devnode and led_get_slot_by_slot_id- for slot_id controller is know so it will always return one slot.

We already have the following functions, we can rename them.

```c
LED_SYM_PUBLIC struct led_slot_list_entry *led_slot_find_by_slot(struct led_ctx *ctx,
								enum led_cntrl_type cntrl,
								char *slot_id);

LED_SYM_PUBLIC struct led_slot_list_entry *led_slot_find_by_device_name(struct led_ctx *ctx,
							enum led_cntrl_type cntrl,
							char *device_name);
```

>It can also apply to set_slot, for example we can impose locate to every controller which supports this devnode using the flag. What do you think?

Please clarify, I'm not exactly following what you mean.  This is the library slot set interface we have today.

```c
led_status_t LED_SYM_PUBLIC led_slot_set(struct led_ctx *ctx, struct led_slot_list_entry *se,
					enum led_ibpi_pattern state);
```

I'm advocating that we add a parameter to `led_slots_get` which defaults to returning what a normal user would want.  Slots that you can call the other functions on and they simply work.  This would happen by only returning slots which are functional.

> Excellent, but I didn't check if ledmon uses it, please verify.

There are no users of `led_get` except the unit tests to test it.

>  know, it is not supported right now. We should consider multiple patterns support in the future. Maybe we should return list instead on enum for get_slot method too?

Could we add a new enum to enum led_ibpi_pattern that indicates both locate and fault are on?

Reply:
> @mtkaczyk
> 
> Today the library API has a call to retrieve the slots:
> 
> ```c
> led_status_t LED_SYM_PUBLIC led_slots_get(struct led_ctx *ctx, struct led_slot_list **sl);
> ```
> 
> You simply get all of them. Ledctl is the one filtering. I'm suggesting that we filter the list in the default case (add argument to change behavior) so that a user never gets the VMD slots or any slots returned to them that they cannot use. At the moment I don't know how to provide that filtering as I don't understand the hardware and when it would work and when it wouldn't. Can you explain this a bit more?

I have a setup with VMD and NPEM on, and device connected:
```
# ledctl -L
{...}
/sys/devices/pci0000:17/0000:17:05.5/pci10000:00/10000:00:00.0/10000:01:00.0/10000:02:06.0 (NPEM)
/sys/devices/pci0000:17/0000:17:05.5 (VMD)
{...}
/sys/devices/pci0000:17/0000:17:05.5/pci10000:00/10000:00:00.0/10000:01:00.0/10000:02:06.0/10000:09:00.0/nvme/nvme3/nvme3n1
```
NPEM is located in downstream port, so the controller path is a path to downstream port. It always exists, even if there is no device connected. The most important is that this is unique for the one drive, there is no problem in obtaining it.

VMD controller is a path to VMD. It is aways available but it is generic device shared between multiple drives. the led management comes with `pciehp` driver in this case, we need to find proper `/sys/bus/pci/slots/<slot>` entry. Address is matching. In this case it is a slot 7:
```
# cat /sys/bus/pci/slots/7/address
10000:09:00
```
Address is a substring on nvme device ( I added ".0"):
```
lspci -s 10000:09:00.0
10000:09:00.0 Non-Volatile memory controller: Intel Corporation NVMe Datacenter SSD [3DNAND, Beta Rock Controller]
```
It doesn't exist if there is no device connected. So the problem is that whiteout disk connected we cannot compare VMD and NPEM slot to determine if they point to the same LED.

We cannot globally assume that if NPEM is behind VMD, then every LED behind VMD is handled by NPEM. There could be 2 PCI switches connected and one may not support NPEM.
That is why I wanted to keep slot interface simple, to be used for more experienced users only and accept duplication here.
Hope it helps!

Reply:
> Not exactly sure what you're referring too. Today the library interface returns the following information about **each** slot:
> 
> * The device node, if it exists
> * The slot identifer
> * The controller
> * The led_ibpi_pattern
> 

> > It applies to set_slot by devnode so probably we will need to make 2 function led_get_slot_by_devnode and led_get_slot_by_slot_id- for slot_id controller is know so it will always return one slot.
> 
> We already have the following functions, we can rename them.
> 
> ```c
> LED_SYM_PUBLIC struct led_slot_list_entry *led_slot_find_by_slot(struct led_ctx *ctx,
> 								enum led_cntrl_type cntrl,
> 								char *slot_id);
> 
> LED_SYM_PUBLIC struct led_slot_list_entry *led_slot_find_by_device_name(struct led_ctx *ctx,
> 							enum led_cntrl_type cntrl,
> 							char *device_name);
> ```
> 
> > It can also apply to set_slot, for example we can impose locate to every controller which supports this devnode using the flag. What do you think?
> 
> Please clarify, I'm not exactly following what you mean. This is the library slot set interface we have today.
> 
> ```c
> led_status_t LED_SYM_PUBLIC led_slot_set(struct led_ctx *ctx, struct led_slot_list_entry *se,
> 					enum led_ibpi_pattern state);
> ```
> 
Apologizes, my brain corrupted at evening... I thought that the interface allows only to set/get the slot for one device/slot not all elements. I didn't check.
Questions have no sense, please ignore.

Reply:
> > I'm advocating that we add a parameter to `led_slots_get` which defaults to returning what a normal user would want. Slots that you can call the other functions on and they simply work. This would happen by only returning slots which are functional.
> 
> Understand and I provided here [#115 (comment)](https://github.com/intel/ledmon/pull/115#discussion_r1281739472) why it is not so simple for VMD and NPEM conflict. We cannot flliter slots if there is no device connected :(

Thanks for the additional information about the VMD and NPEM paths.  I have a some more questions :-)

In comment https://github.com/intel/ledmon/pull/115#discussion_r1255568979 you show output from VMD and NPEM.  In the output you provided, there seems to be a mismatch in the LED state for devices that do correlate, eg.

```
slot: 7               led state: NORMAL          device: /dev/nvme3n1
slot: 10000:02:06.0   led state: FAILURE         device: /dev/nvme3n1
```
Is this because the LED state was changed between each of the queries or because there is a bug in the different paths?  If I'm understanding this correctly, both paths should work right?   What exactly is the `"conflict"`?  Also I'm also assuming that the you didn't include all the output fro the npem case 14 vs 8 devices?

For VMD and NPEM, if the device doesn't exist, there is no LED to change correct?


> > > know, it is not supported right now. We should consider multiple patterns support in the future. Maybe we should return list instead on enum for get_slot method too?
> > 
> > 
> > Could we add a new enum to enum led_ibpi_pattern that indicates both locate and fault are on?
> 
> If you need that- we have to. There is no other option right now. We can handle it in separate pull.

I think this is SES only, right?  I will add a commit to PR which adds this enumeration and sets it for SES.

Reply:
> @mtkaczyk
> 
> > > I'm advocating that we add a parameter to `led_slots_get` which defaults to returning what a normal user would want. Slots that you can call the other functions on and they simply work. This would happen by only returning slots which are functional.
> > 
> > 
> > Understand and I provided here [#115 (comment)](https://github.com/intel/ledmon/pull/115#discussion_r1281739472) why it is not so simple for VMD and NPEM conflict. We cannot flliter slots if there is no device connected :(
> 
> Thanks for the additional information about the VMD and NPEM paths. I have a some more questions :-)
> 
> In comment [#115 (comment)](https://github.com/intel/ledmon/pull/115#discussion_r1255568979) you show output from VMD and NPEM. In the output you provided, there seems to be a mismatch in the LED state for devices that do correlate, eg.
> 
> ```
> slot: 7               led state: NORMAL          device: /dev/nvme3n1
> slot: 10000:02:06.0   led state: FAILURE         device: /dev/nvme3n1
> ```
> 
> Is this because the LED state was changed between each of the queries or because there is a bug in the different paths? If I'm understanding this correctly, both paths should work right? What exactly is the `"conflict"`? Also I'm also assuming that the you didn't include all the output fro the npem case 14 vs 8 devices?

There is one way communication, you can impose the request only. There is no read-back from LED. What is printed to user (using for example get slot) is a cached memory somewhere (it is irrelevant if it is done in kernel or in hardware).
If you are using only one controller all is consistent but if you are mixing ways (like NPEM and VMD does), the one controller is not aware of concurrent update. Here scenario:
```
# reboot - all led cleared
# ledctl --set-slot -c npem -d /dev/nvme0n1 --state=failure
(VMD reports nothing (clear), NPEM reports failure, failure is blinking)
# ledctl --set-slot -c vmd -d /dev/nvme0n1 --state=locate
(VMD reports locate, NPEM reports failure, locate is blinking)
# ledctl --set-slot -c vmd -d /dev/nvme0n1 --state=normal
(VMD reports locate, NPEM reports normal, locate is normal)
```
But as I said now both ways are well respected on Intel hardware but it could change in the future. NPEM is better solution than VMD so VMD may naturally gone.
> 
> For VMD and NPEM, if the device doesn't exist, there is no LED to change correct?

No, LED is always there. https://github.com/intel/ledmon/pull/115#issuecomment-1375671252 - this comment I made about almost a year ago but I think that it is still valid (expect hard requirement to specify controller everytime as we discussed it can be lowered in the future)

You can change LED even if there is no device. That is the capability provided by slot management.
> 
> In the API we have:
> 
> ```c
> LED_SYM_PUBLIC struct led_slot_list_entry *led_slot_find_by_device_name(struct led_ctx *ctx,
> 							enum led_cntrl_type cntrl,
> 							char *device_name);
> ```
> 
> This might be a problem if a device has both VMD & NPEM as we can only return 1 slot id or maybe we're OK as we return the preferred slot which contains the controller?
We are ok because ctrl is mandatory. It is on user now, he chooses.

In the future we can make ctrl field NULL capable and then NPEM should be preferred. that will be simple, just add NPEM controllers first to slot lost.
> 
> > > > know, it is not supported right now. We should consider multiple patterns support in the future. Maybe we should return list instead on enum for get_slot method too?
> > > 
> > > 
> > > Could we add a new enum to enum led_ibpi_pattern that indicates both locate and fault are on?
> > 
> > 
> > If you need that- we have to. There is no other option right now. We can handle it in separate pull.
> 
> I think this is SES only, right? ~I will add a commit to PR which adds this enumeration and sets it for SES.~ I'll place the change in a separate PR.

I don't know. I know well only a few standards here. For sure it doesn't not apply to VMD and NPEM.

Reply:
> No, LED is always there. https://github.com/intel/ledmon/pull/115#issuecomment-1375671252 - this comment I made about almost a year ago

Sorry, I didn't remember this, it has been a while.

> But as I said now both ways are well respected on Intel hardware but it could change in the future. NPEM is better solution than VMD so VMD may naturally gone.

Thanks for explaining this more, it helps.  Now that I think about it we have the same problem with SES and mult-ipath.  More than 1 path to access a LED in a SES enclosure.  It also has the same problem that if you set on one path, you can't retrieve the new status from the other path and visa-versa.

I'm going to think about this some more.  Maybe what we have is fine or maybe we do what you suggested and return a controller enum for `led_is_management_supported`.

> In the future we can make ctrl field NULL capable and then NPEM should be preferred. that will be simple, just add NPEM controllers first to slot lost.

The slot function to retrieve the cntrl field is an enum, can't be null.  We could add a field that indicates some other value.  We do have `LED_CNTRL_TYPE_UNKNOWN` currently.  Adding values to enums is kind of a fuzzy area too, as adding an enum **could** potentially break API users too if they aren't careful.  However, everything could change as we need as we haven't released a `stable` API yet.

Reply:
> I'm going to think about this some more. Maybe what we have is fine or maybe we do what you suggested and return a controller enum for `led_is_management_supported`.

Yes please think. For now the most important is to make the test passing, I would like to finally merge this one.
 I will give us some time for fixes/ changes like this one before I will make new release. Expect your case, I'm not familiar of any hot interest around the library so we should be safe here.

> The slot function to retrieve the cntrl field is an enum, can't be null. We could add a field that indicates some other value. We do have `LED_CNTRL_TYPE_UNKNOWN` currently. Adding values to enums is kind of a fuzzy area too, as adding an enum **could** potentially break API users too if they aren't careful. However, everything could change as we need as we haven't released a `stable` API yet.

Yes, of course it is enum, sorry.
The change like that to allow `LED_CNTRL_TYPE_UNKNOWN` may also impact ledctl- there is a decision to make, perhaps ledctl should allow it too but it is an API change. As you said, before the release we are free to make changes.

Reply:
@mtkaczyk 

I've updated PR

* Removed `led_get`
* Changed return value of `led_is_management_supported` to enum
* Updated the C unit test

Please try this out on your hardware, thanks!

Reply:
It works for me.
```
[gklab-localhost ~/mtkaczyk_work/ledmon/ledmon_lib_pr/tests]# ./runtests.sh
We are running from test folder directly.
exercising ledctl
====================================== test session starts ======================================
platform linux -- Python 3.9.10, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
rootdir: /root/mtkaczyk_work/ledmon/ledmon_lib_pr, configfile: pytest.ini
collecting ...
-------------------------------------- live log collection --------------------------------------
INFO     conftest:conftest.py:11 [CONFIG] ledctl binary: src/ledctl/ledctl
INFO     conftest:conftest.py:12 [CONFIG] slot filters:
collected 2 items

tests/slot_test.py::TestSlot::test_non_slot_set_path[src/ledctl/ledctl-slot_filters0] PASSED [ 50%]
tests/slot_test.py::TestSlot::test_slot_state_walk[src/ledctl/ledctl-slot_filters0] PASSED [100%]

======================================= 2 passed in 0.44s =======================================
running library unit test
Running suite(s): ledmon
100%: Checks: 5, Failures: 0, Errors: 0
```
I will check with @bkucman if we are good to merge it soon.
Thanks!